### PR TITLE
nixos/oci-containers: don't login if image exists locally

### LINKED ISSUE #287673: `virtualisation.oci-containers.containers.<name>.login` fails when the registry is down or you're offline
### Describe the bug
I need to have a container that requires logging into the registry to pull the image. So I setup `virtualisation.oci-container.containers.<name>.login`.

It works fine. But when the registry is down or when I'm offline, restarting the `systemd` service will fail because I failed the `ExecStartPre` step which is the step where the `docker login` command is called. This happened to me when I restarted my computer and my DNS server was down (`network-online.target` is reported to be started even without DNS resolution).

I think the `ExecStartPre` should check if the image exists locally before trying to login. First, this solves my problem as described above, Second, this prevent unneeded logging into the registry because I don't think there's any other reason to log into the registry when you are not pulling image from the registry.

### Steps To Reproduce
1. Have your DNS nameserver to down, maybe make it resolves to something else.
2. Restart the `<docker|podman>-<container-name>.service`

It will fail 5 times and stop trying because of `systemd`

### Expected behavior
I think the better approach is login to the registry **only when** the image is available locally. Maybe change this:
https://github.com/NixOS/nixpkgs/blob/f8e2ebd66d097614d51a56a755450d4ae1632df1/nixos/modules/virtualisation/oci-containers.nix#L254-L259

To something like this:
```
        ${optionalString (isValidLogin container.login) ''
          ${cfg.backend} image inspect ${cfg.image} >/dev/null ||
          ${cfg.backend} login \
          ${container.login.registry} \
          --username ${container.login.username} \
          --password-stdin < ${container.login.passwordFile}
        ''}
```

### Notify maintainers

@adisbladis @lf- @gador 

### Metadata
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 6.1.76, NixOS, 24.05 (Uakari), 24.05.20240207.f8e2ebd`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.18.1`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
```

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

### PR DESCRIPTION
**Pull request by budimanjojo**
Fixes: https://github.com/NixOS/nixpkgs/issues/287673
## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by budimanjojo** [comment]
Tagging @pbsds as the last one that committed to this module before

**Comment by pbsds** [comment]
Can you change the order? Make it so the registry login is always attempted, but the _fallback_ is to check if the image is cached. [Some like to provide `--pull=always`/`newer`](https://github.com/NixOS/nixpkgs/issues/172765#issuecomment-1176954502) to the backend, which then might fail.

**Comment by pbsds** [comment]
please also add a comment as to _why_ that line is there

**Comment by budimanjojo** [comment]
@pbsds My original problem (the issue I link in this PR) is that I don't want the service to fail when it can't login to the registry, which can happen when the registry is down or the DNS is not up yet (not sure why my system reached `network-online.target` before DNS is up).

Maybe I should add another option? Something like `virtualisation.oci-containers.containers.<name>.login.alwaysLogin` as `boolean`?

Please let me know your thoughts so I can update this PR, thanks before.

**Comment by budimanjojo** [comment]
Ah I misunderstood you! What you're saying is changing the order or the check and the login command because people might use `--pull=always` flag.
I can do that and add the comment, but do you think it's better to make this an option like my comment above?

**Comment by budimanjojo** [comment]
I have updated the PR to the suggestion

File: nixos/modules/virtualisation/oci-containers.nix [hunk 1]
```diff
@@ -255,7 +255,9 @@ let
           ${cfg.backend} login \
           ${container.login.registry} \
           --username ${container.login.username} \
-          --password-stdin < ${container.login.passwordFile}
+          --password-stdin < ${container.login.passwordFile} \
+          || ${cfg.backend} image inspect ${container.image} >/dev/null \ # check if image exist locally
+          || echo "image doesn't exist locally and login failed" && exit 1
```
Comment (pbsds) [inline nixos/modules/virtualisation/oci-containers.nix:260]:
```suggestion
          || { echo "image doesn't exist locally and login failed"; exit 1; }
```

Comment (pbsds) [reply nixos/modules/virtualisation/oci-containers.nix:260]: we don't want a `exit 1` if the login succeeds

Comment (budimanjojo) [reply nixos/modules/virtualisation/oci-containers.nix]: my bad, updated

**Review by pbsds — APPROVED** [review]
LGTM, would love a second pair of eyes

**Comment by pbsds** [comment]
> do you think it's better to make this an option like my comment above?

I'm not sure about that option. It's incredibly niche

File: nixos/modules/virtualisation/oci-containers.nix [hunk 2]
```diff
@@ -255,7 +255,9 @@ let
           ${cfg.backend} login \
           ${container.login.registry} \
           --username ${container.login.username} \
-          --password-stdin < ${container.login.passwordFile}
+          --password-stdin < ${container.login.passwordFile} \
+          || ${cfg.backend} image inspect ${container.image} >/dev/null \ # check if image exist locally
```
Comment (SuperSandro2000) [inline nixos/modules/virtualisation/oci-containers.nix:259]: I am somewhat sure that shellcheck would complain about this, as I think the bash ordering is not as we would expect it. Probably some more {} are required maybe after the first || but I am not sure.

Comment (pbsds) [reply nixos/modules/virtualisation/oci-containers.nix:259]: good catch! a `\` before a comment shouldn't work.
Consider moving the comment to above `${cfg.backend}login`

Comment (budimanjojo) [reply nixos/modules/virtualisation/oci-containers.nix]: thanks! updated

**Comment by budimanjojo** [comment]
@Mindavi @h7x4 @SuperSandro2000 @pbsds 

Is everything okay now?

**Pull request merged**